src/wagtail_vector_index/storage/django.py

```diff
@@ -0,0 +1,610 @@
+import logging
+from collections import defaultdict
+from collections.abc import (
+    AsyncGenerator,
+    Generator,
+    Iterable,
+    MutableSequence,
+    Sequence,
+)
+from itertools import chain, islice
+from typing import (
+    TYPE_CHECKING,
+    ClassVar,
+    Optional,
+    Type,
+    TypeAlias,
+    cast,
+)
+
+from django.apps import apps
+from django.core import checks
+from django.core.exceptions import FieldDoesNotExist
+from django.db import models, transaction
+from django.utils.functional import classproperty  # type: ignore
+from wagtail.models import Page
+from wagtail.query import PageQuerySet
+from wagtail.search.index import BaseField
+
+from wagtail_vector_index.ai_utils.backends.base import BaseEmbeddingBackend
+from wagtail_vector_index.ai_utils.text_splitting.langchain import (
+    LangchainRecursiveCharacterTextSplitter,
+)
+from wagtail_vector_index.ai_utils.text_splitting.naive import (
+    NaiveTextSplitterCalculator,
+)
+from wagtail_vector_index.ai_utils.types import TextSplitterProtocol
+from wagtail_vector_index.storage import get_storage_provider, registry
+from wagtail_vector_index.storage.base import (
+    DocumentConverter,
+    DocumentRetrievalVectorIndexMixinProtocol,
+    FromDocumentOperator,
+    ObjectChunkerOperator,
+    ToDocumentOperator,
+    VectorIndex,
+)
+from wagtail_vector_index.storage.exceptions import IndexedTypeFromDocumentError
+from wagtail_vector_index.storage.models import Document
+
+logger = logging.getLogger(__name__)
+
+""" Everything related to indexing Django models is in this file.
+
+This includes:
+
+- The Embedding Django model, which is used to store embeddings for model instances in the database
+- The EmbeddableFieldsMixin, which is a mixin for Django models that lets user define which fields should be used to generate embeddings
+- The EmbeddableFieldsVectorIndexMixin, which is a VectorIndex mixin that expects EmbeddableFieldsMixin models
+- The EmbeddableFieldsDocumentConverter, which is a DocumentConverter that knows how to convert a model instance using the EmbeddableFieldsMixin protocol to and from a Document
+"""
+
+ModelLabel: TypeAlias = str
+ObjectId: TypeAlias = str
+
+
+def batched(iterable, n):
```

Comment:
drive-by suggestion: test if `itertools` has the `batched` attribute (Python 3.12+) and use that before using our implementation. https://docs.python.org/3.12/library/itertools.html#itertools.batched

Reply:
Thanks @zerolab - good call!